homework/philosopher.cpp

```diff
@@ -0,0 +1,144 @@
+#include <algorithm>    // std::for_each()
+#include <atomic>
+#include <chrono>
+#include <cstdlib>
+#include <ctime>
+#include <functional>   // std::mem_fn() 
+#include <iostream>
+#include <mutex>
+#include <string>
+#include <sstream>
+#include <thread>
+
+constexpr size_t itemsCnt = 5;       // number of items (philosophers, forks)
+constexpr size_t actionCnt = 10;     // number of action to take by philosopher
+constexpr bool action_debug = false; // additional debugs
+
+std::mutex forks[ itemsCnt ];
+std::thread philosopher[itemsCnt];
+
+std::mutex displayMutex;
+std::mutex updateMutex;
+std::atomic<bool> canEat[itemsCnt];
+std::atomic<bool> eat[itemsCnt];
+
+using namespace std::chrono_literals;
+
+void displayState()
+{
+    std::stringstream ss;
+
+    std::lock_guard<std::mutex> lock( displayMutex );
+
+    for ( int i = 0; i < itemsCnt; ++i )
+    {
+        if ( eat[i] )
+        {
+            ss << "-F" << i << "- ";
+        }
+        else
+        {
+            ss << " F" << i << "  ";
+        }
+    }
+    ss << std::endl;
+    std::cout << ss.str();
+}
+
+void action( int idx )
+{
+    int left_fork = idx;
+    int right_fork = ( left_fork + 1 ) % 5;
+    int iActionStep = 0;
+
+    // std::cout << "action idx " << idx << " l " << left_fork << " r " << right_fork << std::endl; 
+
+    while ( iActionStep < actionCnt )
+    {
+        if (canEat[idx] == true)
+        {
+            std::scoped_lock lock( forks[left_fork], forks[right_fork] );
+            
+            // NOTE: Here is a race condition on eat[idx] variable and display_state.
+            // Race can be solved by using i.e. queue.
```

Comment:
Tak, jest to problematyczne. Możesz zrobić kolejkę, która będzie przechowywać poszczególne stany programu a dopiero na sam koniec gdy wszystko się wykona wyświetlać całą zawartość kolejki. Możesz też użyć condition_variable i powiadamiać wątek wypisujący, gdy pojawi się nowa rzecz w kolejce.

Reply:
Następna wersja rozwiąże ten problem ...